// fontkit/src/tables/JSTF.js
// Generated by CoffeeScript 1.8.0
(function() {
  var ClassDef, Coverage, Device, FeatureList, GPOSLookup, JstfGSUBModList, JstfLangSys, JstfLangSysRecord, JstfMax, JstfPriority, JstfScript, JstfScriptRecord, LookupList, ScriptList, r, _ref;

  r = require('restructure');

  _ref = require('./opentype'), ScriptList = _ref.ScriptList, FeatureList = _ref.FeatureList, LookupList = _ref.LookupList, Coverage = _ref.Coverage, ClassDef = _ref.ClassDef, Device = _ref.Device;

  GPOSLookup = require('./GPOS').GPOSLookup;

  JstfMax = new r.Array(new r.Pointer(r.uint16, GPOSLookup));

  JstfGSUBModList = new r.Array(r.uint16, r.uint16);

  JstfPriority = new r.Struct({
    shrinkageEnableGSUB: new r.Pointer(r.uint16, JstfGSUBModList),
    shrinkageDisableGSUB: new r.Pointer(r.uint16, JstfGSUBModList),
    shrinkageEnableGPOS: new r.Pointer(r.uint16, JstfGSUBModList),
    shrinkageDisableGPOS: new r.Pointer(r.uint16, JstfGSUBModList),
    shrinkageJstfMax: new r.Pointer(r.uint16, JstfMax),
    extensionEnableGSUB: new r.Pointer(r.uint16, JstfGSUBModList),
    extensionDisableGSUB: new r.Pointer(r.uint16, JstfGSUBModList),
    extensionEnableGPOS: new r.Pointer(r.uint16, JstfGSUBModList),
    extensionDisableGPOS: new r.Pointer(r.uint16, JstfGSUBModList),
    extensionJstfMax: new r.Pointer(r.uint16, JstfMax)
  });

  JstfLangSys = new r.Array(new r.Pointer(r.uint16, JstfPriority), r.uint16);

  JstfLangSysRecord = new r.Struct({
    tag: new r.String(4),
    jstfLangSys: new r.Pointer(r.uint16, JstfLangSys)
  });

  JstfScript = new r.Struct({
    extenderGlyphs: new r.Pointer(r.uint16, new r.Array(r.uint16, r.uint16)),
    defaultJstfLangSys: new r.Pointer(r.uint16, JstfLangSys),
    jstfLangSysCount: r.uint16,
    jstfLangSysRecords: new r.Array(JstfLangSysRecord, 'jstfLangSysCount')
  });

  JstfScriptRecord = new r.Struct({
    tag: new r.String(4),
    jstfScript: new r.Pointer(r.uint16, JstfScript, {
      type: 'parent'
    })
  });

  module.exports = new r.Struct({
    version: r.uint32,
    jstfScriptCount: r.uint16,
    jstfScriptRecords: new r.Array(JstfScriptRecord, 'jstfScriptCount')
  });

}).call(this);
