// fontkit/src/tables/kern.js
// Generated by CoffeeScript 1.8.0
(function() {
  var KernPair, KernSubtable, KernTable, r;

  r = require('restructure');

  KernPair = new r.Struct({
    left: r.uint16,
    right: r.uint16,
    value: r.int16
  });

  KernSubtable = new r.VersionedStruct('format', {
    0: {
      nPairs: r.uint16,
      searchRange: r.uint16,
      entrySelector: r.uint16,
      rangeShift: r.uint16,
      pairs: new r.Array(KernPair, 'nPairs')
    },
    3: {
      glyphCount: r.uint16,
      kernValueCount: r.uint8,
      leftClassCount: r.uint8,
      rightClassCount: r.uint8,
      flags: r.uint8,
      kernValue: new r.Array(r.int16, 'kernValueCount'),
      leftClass: new r.Array(r.uint8, 'glyphCount'),
      rightClass: new r.Array(r.uint8, 'glyphCount'),
      kernIndex: new r.Array(r.uint8, function() {
        return this.leftClassCount * this.rightClassCount;
      })
    }
  });

  KernTable = new r.VersionedStruct('version', {
    0: {
      subVersion: r.uint16,
      length: r.uint16,
      format: r.uint8,
      coverage: new r.Bitfield(r.uint8, ['horizontal', 'minimum', 'crossStream', 'override']),
      subtable: KernSubtable,
      padding: new r.Reserved(r.uint8, function() {
        return this.length - this._currentOffset;
      })
    },
    1: {
      length: r.uint32,
      coverage: new r.Bitfield(r.uint8, [null, null, null, null, null, 'variation', 'crossStream', 'vertical']),
      format: r.uint8,
      tupleIndex: r.uint16,
      subtable: KernSubtable,
      padding: new r.Reserved(r.uint8, function() {
        return this.length - this._currentOffset;
      })
    }
  });

  module.exports = new r.VersionedStruct(r.uint16, {
    0: {
      nTables: r.uint16,
      tables: new r.Array(KernTable, 'nTables')
    },
    1: {
      reserved: new r.Reserved(r.uint16),
      nTables: r.uint32,
      tables: new r.Array(KernTable, 'nTables')
    }
  });

}).call(this);
