// fontkit/src/glyph/SBIXGlyph.js
// Generated by CoffeeScript 1.8.0
(function() {
  var SBIXGlyph, TTFGlyph, r,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  TTFGlyph = require('./TTFGlyph');

  r = require('restructure');

  SBIXGlyph = (function(_super) {
    var SBIXImage;

    __extends(SBIXGlyph, _super);

    function SBIXGlyph() {
      return SBIXGlyph.__super__.constructor.apply(this, arguments);
    }

    SBIXImage = new r.Struct({
      originX: r.uint16,
      originY: r.uint16,
      type: new r.String(4),
      data: new r.Buffer(function() {
        return this.parent.buflen - this._currentOffset;
      })
    });

    SBIXGlyph.prototype.getImageForSize = function(size) {
      var end, offsets, start, table, _i, _len, _ref;
      _ref = this._font.sbix.imageTables;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        table = _ref[_i];
        if (table.ppem >= size) {
          break;
        }
      }
      offsets = table.imageOffsets;
      start = offsets[this.id];
      end = offsets[this.id + 1];
      if (start === end) {
        return null;
      }
      this._font.stream.pos = start;
      return SBIXImage.decode(this._font.stream, {
        buflen: end - start
      });
    };

    SBIXGlyph.prototype.render = function(ctx, size) {
      var img, scale;
      img = this.getImageForSize(size);
      if (img != null) {
        scale = 1 / this._font.head.unitsPerEm * size;
        ctx.image(img.data, {
          height: size,
          x: img.originX,
          y: (this.bbox[1] - img.originY) * scale
        });
      }
      if (this._font.sbix.flags.renderOutlines) {
        return SBIXGlyph.__super__.render.apply(this, arguments);
      }
    };

    return SBIXGlyph;

  })(TTFGlyph);

  module.exports = SBIXGlyph;

}).call(this);
